### PR TITLE
streamingccl,sql: add crdb_internal.unsafe_revert_tenant_to_timestamp

### PR DESCRIPTION
**Pull request by stevendanna**
This function reverts the given tenant to the given timestamp. This is
a destructive, non-transactional operation.

This should only be used with extreme caution.

It's use currently requires:

- a user with ADMIN or MANAGETENANT privileges,
- a session where sql_safe_updates=false, and
- a target tenant with a service mode of None.

While the code attempts to install a session-bound PTS record on the
revert timestamp, if the given timestamp has already been garbage
collected, it is possible that the data has been partially reverted
before an error is returned. As such, extreme care should be take to
ensure that this is only called with a timestamp that is known to be
under the gcThreshold of all relevant ranges.

Epic: none

Release note: None

### CODE REVIEW

**Comment by stevendanna** [comment]
The first commit is a small rename and metadata addition. See it's commit message for details. I can split it into another PR if necessary.

File: pkg/multitenant/mtinfopb/info.proto [hunk 1]
```diff
@@ -49,11 +50,17 @@ message ProtoInfo {
...
+  // PhysicalReplicationConsumerJobID is set if this tenant is the target tenant
   // of a running tenant replication job.
-  optional int64 tenant_replication_job_id = 4 [
+  optional int64 physical_replication_consumer_job_id = 4 [
      (gogoproto.nullable) = false,
-     (gogoproto.customname) = "TenantReplicationJobID",
+     (gogoproto.customname) = "PhysicalReplicationConsumerJobID",
+     (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb.JobID"];
+
+  // PhysicalReplicationConsumerJobIDs contains any physical
```
Comment (adityamaru) [inline pkg/multitenant/mtinfopb/info.proto:60]: s/PhysicalReplicationConsumerJobIDs/PhysicalReplicationProducerJobIDs/g

File: pkg/multitenant/mtinfopb/info.proto [hunk 2]
```diff
@@ -62,9 +69,30 @@ message ProtoInfo {
     (gogoproto.nullable) = false
   ];
 
-  // Next ID: 7
+  // If this tenant was previous a physical replication target that
```
Comment (adityamaru) [inline pkg/multitenant/mtinfopb/info.proto:72]: s/previous/previously/g

File: pkg/multitenant/mtinfopb/info.proto [hunk 3]
```diff
@@ -62,9 +69,30 @@ message ProtoInfo {
...
+  // was promoted. This is the timestamp at which it was promoted.
+  optional PreviousSourceTenant previous_source_tenant = 7 [(gogoproto.nullable) = false];
```
Comment (adityamaru) [inline pkg/multitenant/mtinfopb/info.proto:74]: I'm probably thinking about this the wrong way but `PreviousSourceTenant` to me implies that this is information about when this tenant was the source of the replication stream. The comment however talks about this being the target (i.e. destination) and when it was promoted to be the primary after cutover. I think in this case the comment is correct but the name is a little confusing?

File: pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go [hunk 4]
```diff
@@ -52,14 +59,124 @@ func (r *streamIngestManagerImpl) GetStreamIngestionStats(
 	return replicationutils.GetStreamIngestionStats(ctx, streamIngestionDetails, jobProgress)
 }
 
+// GetStreamIngestionStatsAndStatus implements streaming.StreamIngestManager interface.
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go:62]: s/GetStreamIngestionStatsAndStatus/GetReplicationStatsAndStatus/g

File: pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go [hunk 5]
```diff
@@ -52,14 +59,124 @@ func (r *streamIngestManagerImpl) GetStreamIngestionStats(
...
+		[]roachpb.Span{spanToRevert},
+		revertTo,
+		false, /* ignoreGCThreshold */
+		int64(sql.RevertTableDefaultBatchSize),
+		nil /* onCompletedCallback */); err != nil {
+		return err
+	}
+
+	return execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantRecord, err := sql.GetTenantRecordByName(ctx, r.evalCtx.Settings, r.txn, tenantName)
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go:137]: you're passing in r.txn here, I think you meant to pass in txn instead?

File: pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go [hunk 6]
```diff
@@ -52,14 +59,124 @@ func (r *streamIngestManagerImpl) GetStreamIngestionStats(
...
+		if err != nil {
+			return err
+		}
+		tenantRecord.DataState = originalDataState
+		return sql.UpdateTenantRecord(ctx, r.evalCtx.Settings, r.txn, tenantRecord)
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go:142]: ditto as above

**Review by adityamaru — CHANGES_REQUESTED** [review]
really slick!

File: pkg/multitenant/mtinfopb/info.proto (hunk 3 above)
Comment (stevendanna) [reply pkg/multitenant/mtinfopb/info.proto:74]: Right, the tenant the overall ProtoInfo describes used to be a destination tenant whose source is the cluster identified by PreviousSourceTenant.  Given that, any ideas about a better name?

File: pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go (hunk 5 above)
Comment (stevendanna) [reply pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go:137]: Good catch, thanks!

File: pkg/multitenant/mtinfopb/info.proto (hunk 3 above)
Comment (stevendanna) [reply pkg/multitenant/mtinfopb/info.proto:74]: (I think this name will make more sense in my next commit that adds the ClusterID and TenantID of the previous source cluster.

Comment (adityamaru) [reply pkg/multitenant/mtinfopb/info.proto:75]: the new comment makes this much clearer, I also see it clearer now when thinking about it in the context of the broader ProtoInfo.

**Review by adityamaru — APPROVED** [review]

File: pkg/sql/catalog/bootstrap/bootstrap_test.go [hunk 7]
```diff
@@ -176,7 +176,7 @@ func TestSystemDatabaseSchemaBootstrapVersionBumped(t *testing.T) {
 
 	// If you need to update this value (i.e. failed this test), check whether
 	// you need to bump systemschema.SystemDatabaseSchemaBootstrapVersion too.
-	const prevSystemHash = "b23cc2e4f5d8e1a3d3bdd514addb3bd9af3eff65cd485fa2822ced1a6a2e8fd4"
+	const prevSystemHash = "b88564c021a413e53ed91a656fd43d3212072070660e3f9e51539bcb7935ec25"
```
Comment (stevendanna) [inline pkg/sql/catalog/bootstrap/bootstrap_test.go:179]: @dt I believe this is fine. But please double check.

File: pkg/sql/logictest/testdata/logic_test/tenant_builtins [hunk 8]
```diff
@@ -4,7 +4,7 @@ query IBIT colnames
 SELECT id, active, length(info), name FROM system.tenants ORDER BY id
 ----
 id  active  length  name
-1   true    10      system
+1   true    14      system
```
Comment (dt) [inline pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: Why did adding the fields to the proto -- but not setting them -- make this go up?

File: pkg/sql/catalog/bootstrap/bootstrap_test.go (hunk 7 above)
Comment (dt) [reply pkg/sql/catalog/bootstrap/bootstrap_test.go:179]: I mean, I think it is "fine" but I'd like to understand why it happened, thus my comment on the extra 4 bytes observed on tenant 1 record.

File: pkg/sql/logictest/testdata/logic_test/tenant_builtins (hunk 8 above)
Comment (stevendanna) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: My 80% sure guess is that this is the result of gogoproto.nullable = false.  While the new fields are "optional", by default we'll be putting zero-valued messages in both the new hlc.Timestamp and the enclosing message.  Both of those likely have 1 byte for the tag and 1 byte for the lengths. 2*(1+1) = 4.

Comment (dt) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: I _thought_ proto didn't encode fields that had their default/zero value?

> Also note that if a scalar message field is set to its default, the value will not be serialized on the wire.

Comment (stevendanna) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: We could remove nullable=false from the top level message and probably get these bytes back in the default case.  We'll have to do a nil-check before accessing that field, but not on any of the fields inside of it which seems like a fine tradeoff.

Comment (dt) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: maybe we should just make the message type nullable; I'm not going to lose sleep over one alloc per tenant?

Comment (stevendanna) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins:7]: > I thought proto didn't encode fields that had their default/zero value?

I have a feeling this is the result of proto2's optional and gogoproto's nullable kinda being at cross purpose.

Comment (stevendanna) [reply pkg/sql/logictest/testdata/logic_test/tenant_builtins]: Nullable it is.

**Review by dt — APPROVED** [review]

**Comment by stevendanna** [comment]
bors r=adityamaru

**Pull request merged**